docs/Overview.md

```diff
@@ -1,96 +1,246 @@
-* Plan Node - A plan node represents a single step in the DAG of a plan.
-* Executor - The executor is responsible for processing a query plan.
-    The executor process data via data frames.
-* Storage - The Storage interface provides a mechanism for accessing the underlying data as data frames.
-* Capabilities - The Storage interface exposes its capabilities.
-    The planner uses the available capabilities to determine the best plan.
-* Hints - The Storage interface exposes hints about the data.
-    The planner uses the hints to determine the best plan.
-* Query Engine - Query Engine is the name given to the entire system being described in this document.
+The engine provides methods for registering sources and transformations, and for binding builtin values used for interpretation.
+Sources and transformations constitute the actual implementation of the data manipulation operations of some function values in the scope.
 
+```go
+type CreateSource = fn (ExecutionContext) Source
+type CreateTransformation = fn (ExecutionContext) Transformation
+type CompilerType interface {
+    From() Language
+    To() Language
+}
 
-Both a query and a plan are represented by a DAG and describe an operation that needs to be performed.
-The difference is that a plan in addition to describing what the operation is, also describes how that operation will be performed.
-In short, a query describes what the operation is and a plan describes how that operation will be carried out.
+type Engine interface {
+    BindValue(id string, Value)
+    Prelude() Scope
+    
+    RegisterSource(id string, CreateSource)
+    RegisterTransformation(id string, CreateTransformation)
+    
+    Query(Representation) (Query, error)
+}
 
-## Relations
+type Scope interface {} // ...
+type Source interface {} // ...
+type Transformation interface {} // ...
+type Value interface {} // ...
+```
 
-Below is a high level description, using the Go language, of the relations between the different components and concepts of the query engine.
+A language is used to describe a query. It consists of words whose letters are taken from an alphabet.
+Letters can be of any type and complexity, from characters, to nodes in an Abstract Syntax Tree (AST). Various languages
+can be used to represent a query.
 
 ```go
-
-type Parser interface {
-    Parse(flux string) Query
+type Language interface {
+    Name() string
 }
+```
 
-// Query defines work to be performed on time series data and a result.
-// A query is represented as a directed acyclic graph (DAG).
+A query defines a unit of work for the engine. It comprises one or more pipelines.
+It contains its representation, its results, and execution statistics.
+
+```go
 type Query interface {
-    Nodes() []QueryNode
-    Edges() []Edge
+    Results() Results
+    Stats() map[string]interface{}
+    Representation() Representation
+}
+
+type Representation interface {
```

Comment:
I think this implementation could be simplified a lot. It's unclear to me why it needs to be a `Visitor` as that seems to be very dependent on the internal structure and related to the compiler itself. I can't find a definition of the `Visitor` interface with a quick glance though.

I also think that `Representation` and `Language` are essentially the same thing and the name is very vague. I think `Language` should be the struct that defines the language content and we should just use strings for the language names. I don't see how an extra `Language` interface adds much.

I also do not think we should add a `String()` method to the representation. When I think of `String()` I think of a small thing I can use with logging and printing. This is guaranteed to not be the case with the representation and I'd like it if we played cautiously and did not include this at first. I think there's a reasonable belief that we will need it eventually, but I'd like to add something rather than remove it at the current moment.

Reply:
Agreed on removing `Visitor` (it was there only for explicitly stating that the Representation can be consulted, but you are right, it makes no sense 😅 ).

Agreed on removing `String`.

I think that `Language` and `Representation` are pretty different concept and should stay separated, they can be considered as different as "type" and "object". Indeed `Language` is used for the types of compilers and executors.